### PR TITLE
in-house cost logic for finer control

### PR DESCRIPTION
**Pull request by oxade**
This PR in-houses our gas control logic, which allows us tune exact bytecode cost logic and natives logic.
This is an NFC, but subsequent PRs will build on this and introduce more control and fine-tuning the constants and natives, then hooking this up to estimator.

### CODE REVIEW

File: crates/sui-cost-tables/src/tables.rs [hunk 1]
```diff
@@ -0,0 +1,1120 @@
...
+            ImmBorrowGlobal(StructDefinitionIndex::new(0)),
+            GasCost::new(23, 1),
+            false,
+        ),
+        (
+            ImmBorrowGlobalGeneric(StructDefInstantiationIndex::new(0)),
+            GasCost::new(14, 1),
+            false,
+        ),
+        // Not supported yet in Move
```
Comment (tnowacki) [inline crates/sui-cost-tables/src/tables.rs:371]: There is no syntax to call it with a non-zero size, but the bytecode instruction is still used

File: crates/sui-cost-tables/src/tables.rs [hunk 2]
```diff
@@ -0,0 +1,1120 @@
...
+            Opcodes::NEQ,
+            lhs.legacy_abstract_memory_size() + rhs.legacy_abstract_memory_size(),
+        )
+    }
+
+    fn charge_load_resource(&mut self, _loaded: Option<NumBytes>) -> PartialVMResult<()> {
+        Ok(())
+    }
+
+    fn charge_borrow_global(
```
Comment (tnowacki) [inline crates/sui-cost-tables/src/tables.rs:709]: Should we just invariant violation for all of these global storage operations?

File: crates/sui-cost-tables/src/tables.rs [hunk 3]
```diff
@@ -0,0 +1,1120 @@
...
+        ),
+        (
+            ImmBorrowFieldGeneric(FieldInstantiationIndex::new(0)),
+            GasCost::new(2, 1),
+            true,
+        ),
+        //
+        // Ref
+        //
+        (WriteRef, GasCost::new(2, 0), true),
```
Comment (tnowacki) [inline crates/sui-cost-tables/src/tables.rs:278]: Does write ref need to be size dependent?

File: crates/sui-cost-tables/src/tables.rs [hunk 4]
```diff
@@ -0,0 +1,1120 @@
...
+        (VecSwap(SignatureIndex::new(0)), GasCost::new(2, 0), true),
+        // TODO: This is actually a linear function (dependent on object size and vec length) with a small base constant but that's not easy to model yet
+        (VecPack(SignatureIndex::new(0), 0), GasCost::new(7, 0), true),
+        // Performs a copy
+        (
+            VecImmBorrow(SignatureIndex::new(0)),
+            GasCost::new(3, 1),
+            true,
+        ),
+        // Performs a copy
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-cost-tables/src/tables.rs:220]: Right MutBorrow does not perform copy.

File: crates/sui-cost-tables/src/tables.rs (hunk 1 above)
Comment (oxade) [reply crates/sui-cost-tables/src/tables.rs:371]: How do I trigger it so I can calibrate it?

File: crates/sui-cost-tables/src/tables.rs [hunk 5]
```diff
@@ -0,0 +1,1120 @@
...
+        ),
+        // Swap depends on size?
+        (
+            VecPushBack(SignatureIndex::new(0)),
+            GasCost::new(3, 0),
+            true,
+        ),
+        // TODO: Do we scale by memory for swap?
+        (VecSwap(SignatureIndex::new(0)), GasCost::new(2, 0), true),
+        // TODO: This is actually a linear function (dependent on object size and vec length) with a small base constant but that's not easy to model yet
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-cost-tables/src/tables.rs:212]: @tnowacki it was not easy to model in the old cost logic which had very limited params. But now it's doable.

**Comment by oxade** [comment]
@tnowacki  since the staging table is still WIP and might be confusing, I decided to just back it out for now while we flesh it out fully. This handles most of your comments

File: crates/sui-cost-tables/src/tables.rs (hunk 2 above)
Comment (oxade) [reply crates/sui-cost-tables/src/tables.rs:457]: Will do. This includes `Exists`, `MoveFrom`, `MoveTo`, right?

Comment (tnowacki) [reply crates/sui-cost-tables/src/tables.rs:457]: Correct

File: crates/sui-cost-tables/src/tables.rs (hunk 1 above)
Comment (oxade) [reply crates/sui-cost-tables/src/tables.rs]: Okay I will add it to the calibration list.

**Review by tnowacki — APPROVED** [review]

**Pull request merged**